pkg/actions/karpenter/tasks.go

```diff
@@ -0,0 +1,142 @@
+package karpenter
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/aws/aws-sdk-go/aws"
+	cfn "github.com/aws/aws-sdk-go/service/cloudformation"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/kris-nova/logger"
+
+	api "github.com/weaveworks/eksctl/pkg/apis/eksctl.io/v1alpha5"
+	"github.com/weaveworks/eksctl/pkg/cfn/builder"
+	"github.com/weaveworks/eksctl/pkg/cfn/manager"
+	"github.com/weaveworks/eksctl/pkg/karpenter"
+	"github.com/weaveworks/eksctl/pkg/utils/tasks"
+)
+
+const (
+	kubernetesTagFormat = "kubernetes.io/cluster/%s"
+)
+
+type karpenterTask struct {
+	info               string
+	stackManager       manager.StackManager
+	cfg                *api.ClusterConfig
+	ec2API             ec2iface.EC2API
+	karpenterInstaller karpenter.InstallKarpenter
+}
+
+func (k *karpenterTask) Describe() string { return k.info }
+func (k *karpenterTask) Do(errs chan error) error {
+	return k.createKarpenterTask(errs)
+}
+
+// NewTasksToInstallKarpenter defines tasks required to create Karpenter
+func NewTasksToInstallKarpenter(cfg *api.ClusterConfig, stackManager manager.StackManager, ec2API ec2iface.EC2API, karpenterInstaller karpenter.InstallKarpenter) *tasks.TaskTree {
+	taskTree := &tasks.TaskTree{Parallel: true}
+	taskTree.Append(&karpenterTask{
+		info:               fmt.Sprintf("create karpenter for stack %q", cfg.Metadata.Name),
+		stackManager:       stackManager,
+		cfg:                cfg,
+		ec2API:             ec2API,
+		karpenterInstaller: karpenterInstaller,
+	})
+	return taskTree
+}
+
+// createKarpenterTask creates Karpenter
+func (k *karpenterTask) createKarpenterTask(errs chan error) error {
+	name := k.makeKarpenterStackName()
+
+	logger.Info("building nodegroup stack %q", name)
+	stack := builder.NewKarpenterResourceSet(k.cfg)
+	if err := stack.AddAllResources(); err != nil {
+		return err
+	}
+	tags := map[string]string{
+		api.KarpenterNameTag: name,
+	}
+	if err := k.stackManager.CreateStack(name, stack, tags, nil, errs); err != nil {
+		return fmt.Errorf("failed to create stack: %w", err)
+	}
+
+	if err := k.maybeUpdateSubnetTags(); err != nil {
+		return err
+	}
+	return k.karpenterInstaller.Install(context.Background())
+}
+
+// makeNodeGroupStackName generates the name of the Karpenter stack identified by its name, isolated by the cluster this StackCollection operates on
+func (k *karpenterTask) makeKarpenterStackName() string {
+	return fmt.Sprintf("eksctl-%s-karpenter", k.cfg.Metadata.Name)
+}
+
+// maybeUpdateSubnetTags will check if the kubernetes.io/cluster tag is present on the subnets.
+// if not, it will create them.
+func (k *karpenterTask) maybeUpdateSubnetTags() error {
+	var ids []string
+	for _, subnet := range k.cfg.VPC.Subnets.Private {
+		ids = append(ids, subnet.ID)
+	}
+	for _, subnet := range k.cfg.VPC.Subnets.Public {
+		ids = append(ids, subnet.ID)
+	}
+	sort.Strings(ids)
```

Comment:
Because they are coming from a map.

Reply:
right but why do they need to be ordered? 🤔

Reply:
not sure what I'm not seeing :')

Reply:
@nikimanoledaki Can I resolve this? :)

Reply:
>  Because the unit test wouldn't be able to match the calling order

annoying! You can get around this by doing something like:

```
                     p.MockEC2().On("DescribeSubnets", mock.Anything)..Run(func(args mock.Arguments) {
                          Expect(args).To(HaveLen(1))
                          Expect(args[0]).To(BeAssignableToTypeOf(&awseks.DescribeSubnetsInput{}))
                          Expect(args[0].(*awseks.DescribeSubnetsInput).Subnets).To(ConsistOf(.....) //unordered matcher
                      }.Return(&ec2.DescribeSubnetsOutput{
				Subnets: []*ec2.Subnet{
					{
						SubnetId: aws.String(privateSubnet1),
						VpcId:    aws.String(cfg.VPC.ID),
					},
					{
						SubnetId: aws.String(privateSubnet2),
						VpcId:    aws.String(cfg.VPC.ID),
						Tags: []*ec2.Tag{
							{
								Key:   aws.String("kubernetes.io/cluster/" + clusterName),
								Value: aws.String("shared"),
							},
						},
					},
					{
						SubnetId: aws.String(publicSubnet1),
						VpcId:    aws.String(cfg.VPC.ID),
					},
					{
						SubnetId: aws.String(publicSubnet2),
						VpcId:    aws.String(cfg.VPC.ID),
					},
				},
			}, nil)
```

>

Reply:
Hmmm, hmmm. I like the self-verifying call order. But I'll consider this. :D

Reply:
I prefer this because I'd rather have a bit more complexity in the test than in the code. Also easy for someone to see the code and go "this is overly complicated, let me remove the sorting logic"

Reply:
Okay, fine. :D I'll add it.

Reply:
This is just horrible, to be honest. :D 

```
                          Expect(args[0].(*awseks.DescribeSubnetsInput).Subnets).To(ConsistOf(.....) //unordered matcher

``

Reply:
Yeah, I hate this. I also have to change all of the CreateTag calls too because they contain the updateSubnetIds which can be in any order. Instead of doing all that in the tests I can simplify the whole thing with a nice and easy sort. Which is what I'm going to do. :) If you truly hate it, you can change it in an append PR. :D I will merge it. :D